### DISCUSSION TITLE
Tauri or Flutter for RustDesk desktop?

### DISCUSSION THREAD

**Discussion by rustdesk** [rustdesk/rustdesk#533] [Polls]
Though our open source [mobile side](https://github.com/rustdesk/rustdesk/tree/master/flutter) is Flutter, but I can not decide either. Our desktop side uses Sciter, similar to html/js. We can use most of old html/js with Tauri. We can also reuse some mobile Flutter code, but our desktop design is super different from mobile. Flutter is not css based, few code can be shared between mobile and desktop per my understanding.

Drawback of flutter:

- Debugging rust with Flutter is not conveninent.

- Both of them are open source, but Tauri has much smaller code base, we can do code change on Tauri, but it is hard for us on Flutter.

----
Our final choice is Flutter

**Comment by fzyzcjy** [comment]
Personally I favor Flutter for its rapid development speed and feature richness. (I wrote >300k lines of code Flutter app, deployed in both mobile Android/iOS and desktop Windows/MacOS/Linux)

> Debugging rust with Flutter is not conveninent.

Flutter can communicate with Rust conveniently via https://github.com/fzyzcjy/flutter_rust_bridge/ (Disclaimer: I wrote that library, originally for my personal use)

> We can also reuse some mobile Flutter code, but desktop is super different from mobile. Flutter is not css based, few code can be shared between mobile and desktop per my understanding.

Not sure about your case, but for my own case it is very similar and I reused almost all code. If you are talking about different UI, flutter's widgets are like React widgets, and you can freely write some conditional branches to combine widgets differently on desktop and mobile.

**Reply by rustdesk** [reply]
> https://github.com/fzyzcjy/flutter_rust_bridge/

I have given you a star, Respect!
How do you debug Rust with Flutter? With Tauri, `cargo run` is convenient for debug. Before I wrote Flutter, I `cargo build` first, then copy `so` file to jni folder, then `flutter run`, it is not friendly.

**Reply by fzyzcjy** [reply]
> How do you debug Rust with Flutter? 

I use logging instead of debuggers. IMHO logging-based debugging is as fast as debugger-based debugging and is even more powerful in many scenarios.

**Reply by rustdesk** [reply]
> logging

I use logging method also, but I miss `cargo run`.

**Reply by fzyzcjy** [reply]
Have a look at the doc [fzyzcjy.github.io/flutter_rust_bridge/](https://fzyzcjy.github.io/flutter_rust_bridge/). With some setup you can only run `flutter run` and it will invoke cargo automatically for you.

**Reply by rustdesk** [reply]
> 

Do you know if  Flutter support command line now? We need the same one exe can work as GUI and as console both.

**Reply by fzyzcjy** [reply]
Sure! Just write a pure-Dart program. I have also done that for my app. Basically you can (1) write some code that is pure Dart, like domain logic (2) write some command line wrapper code, using the code in 1 and possibly https://pub.dev/packages/args (3) write some Flutter-specific code and run it with flutter gui

**Reply by fzyzcjy** [reply]
Btw Flutter even has improved e2e testing utility now: https://github.com/fzyzcjy/flutter_convenient_test (Disclaimer: Again written by me, again used internally heavily and later open sourced). So its testability is as good as web-based solutions.

**Reply by rustdesk** [reply]
> https://github.com/fzyzcjy/flutter_convenient_test

You are brilliant.

**Reply by fzyzcjy** [reply]
Hope you also join Flutter - the more people contributing to Flutter open-source ecosystem (though already quite a lot), 
the better every developer's experience will be. (Btw I feel rustdesk quite interesting and will give it a try when needed)

**Reply by sagudev** [reply]
I was using Tauri before and it was nice, as I know web technologies pretty well, but my app used to send a lot of images to Tauri from Rust and it was pretty slow. I changed from Tauri callbacks to using web-server inside rust app and run Tauri just to display this webapp, which still used some Tauri exposed JavaScript APIs and it was faster but still slow for 1000 images that were loaded in the list. See more on [Tauri discord](https://discord.com/channels/616186924390023171/731495028677148753/929675404086767637).

I decided to port my app to flutter, but as I am learning flutter at the same time I am progressing slowly. In that time I have also authored commits and opened bugs at [flutter_rust_bridge](https://github.com/fzyzcjy/flutter_rust_bridge/). Will report when I will have rough numbers for same scenario using flutter and flutter_rust_bridge.

**Reply by taviks** [reply]
@sagudev 

> Will report when I will have rough numbers for same scenario using flutter and flutter_rust_bridge.

Any updates?

**Reply by davidhmays** [reply]
@sagudev how'd the Flutter version go?

**Reply by sagudev** [reply]
I abandoned project altogether (tauri one too), although it would be interesting to do just small POC to test transfer speeds, but I currently do not have time for this.

**Comment by rustdesk** [comment]
Copy from Reddit, very insightful, https://www.reddit.com/r/rust/comments/uqr1nq/tauri_or_flutter_for_rustdesk_desktop/i8szbvv/?context=3
> Having tried both (and Dioxus too), Tauri seems to me a better choice.
> 
> The release candidate they released is pretty solid, tooling is great (create-tauri-app), communication between frontend and backend is very simple (it's just a macro, it supports async, and everything that implements seeds::(De)Serialize can be sent/received), etc.
> 
> Also, since it uses web technologies, you get all the existing ecosystem, which is huge. It's unbeatable.
> 
> Flutter is great for mobile apps. For desktop, there's still a lot missing. Performance is also very poor (honestly wasn't expecting that, might have something to do with my hardware thought).

**Reply by fzyzcjy** [reply]
> Performance is also very poor (honestly wasn't expecting that, might have something to do with my hardware thought).

Interesting observation! Personally I feel it smooth.

By the way, Flutter 3.0 is released just *a few days ago* with improvements especially on desktop. Not sure whether it has improvements on performance.

Anyway I feel both Flutter and Tauri are excellent :)

**Reply by rustdesk** [reply]
> Flutter 3.0

Wondering if Flutter 3 fixes Canvaskit UTF font issue. :)

**Reply by fzyzcjy** [reply]
@rustdesk Not sure what is that issue? (not heard before) But seems it is closed long ago https://github.com/flutter/flutter/issues/58496

Btw I guess you will use flutter Desktop, not Flutter Web :)

**Comment by shirshak55** [comment]
I am voting Tauri for the following reasons:

- Tauri has huge ecosystem due to html/css/js
- Tauri Community is awesome.
- Tauri is more mature in desktop work compared to flutter.
- With Tauri, it is easier to get more contributors.
- You can utilize the full power of Rust ( Flutter Rust communication is probably easy too. In Tauri it is just a single macro)

**Reply by rustdesk** [reply]
Sadly, Tauri has poor ipc, which is very important for us, we can not afford sacrificing performace by choosing Tauri. I do not think Tauri can solve this problem in short term because it highly depends on underlying webview.

**Reply by tcanabrava** [reply]
what about gtk-rs or qt-rs? there are pure rust alternatives, too, that do not depend on node, or dart.

**Comment by rustdesk** [comment]
https://www.reddit.com/r/rust/comments/uqr1nq/tauri_or_flutter_for_rustdesk_desktop/i8thnbk/?context=3

> Definitely recommend Tauri.
> 
> - reuse (almost) all of your html/css/js, with easy native extensions
> 
> - no need to learn a new language
> 
> - the bundle size is small
> 
> - it's portable without a ton of effort
> 
> For projects like this (with a small base of contributors), the more code you can share, the better, so I would say that's the most important factor.

**Comment by rustdesk** [comment]
https://www.reddit.com/r/FlutterDev/comments/uqr18d/tauri_or_flutter_for_rustdesk_desktop/i8tj87f/?context=3

> I'll second the possibility of sharing UI and code between mobile and desktop. I'm not sure why you think that's not possible with Flutter, it's actually really easy. Overall it seems like your choice is between what's easiest in the short term (Tauri, since your desktop UI is already written in HTML/CSS) versus what's easier to maintain in the long run (Flutter, since you'd only have 1 codebase instead of 2).

**Reply by nothingismagick** [reply]
FYI - Tauri is going to be releasing mobile this year, and will also be providing EGUI windowing options (at least on desktop, although it seems by the time we're ready for Mobile that EGUI will be properly available on mobile).

**Reply by LifeIsStrange** [reply]
> FYI - Tauri is going to be releasing mobile this year, and will also be providing EGUI windowing options (at least on desktop, although it seems by the time we're ready for Mobile that EGUI will be properly available on mobile).

@nothingismagick You can use Ionic in the meantime

**Reply by nothingismagick** [reply]
One could.

**Comment by nothingismagick** [comment]
In addition to the very positive comments for both approaches, what you will also find in the Tauri community, is that people there are very accessible and helpful. Not only does the core team gladly review and accept PRs - we also actively adopt upstream projects in order to maintain them so that Tauri and the broader open-source community can continue growing. But again, both are solid choices. 

Maybe the things that resonated the most, are:
- the time involved in retooling (sciter -> flutter will be a lot of work)
- the ability for you to acquire contributors (flutter dev is a smaller pool of equally talented engineers)

**Comment by NathanaelA** [comment]
I have an entire set of blog articles on Tauri, Flutter and other cross platform technology bases starting here https://fluentreports.com/blog/?p=1288  -- however for those who don't want to read an entire book.  :grinning:      

1. Flutter, Dart based, smaller memory footprint than Tauri -- everything is under control by app on all platforms.  Requires slightly better hardware than Tauri as it needs OpenGL/Vulken/Metal support.
2. Tauri, Mix of rust (which is a sweet integration!) and html/css/JS  - memory footprint almost the same as Electron!   Many bugs are outside of your control since your app will be running on different underlying engine on each _PLATFORM_ and _VERSION_ of the platform (i.e. Ubuntu 18 LTS has a different Webkit version than 20, 21 & 22.  Now think exponentially how many different distro's and versions they all each have... Now include Windows Edge/Blink engine versions, and then also Mac OS Webkit versions...   Throw into that different rendering of Blink vs Webkit and different JS gotchas between JSCore & V8 and you have a wonderful recipe for some fun times... :wink: )

If your only two choices are Flutter and Tauri, imho, Flutter is the significantly (& I seriously mean significantly) safer choice, even though I'm personally not a big fan of Flutter.     As much as I don't like Electron apps, Electron is also a much better choice than Tauri because the Browser is the exact same fixed engine version for all platforms, which then eliminates a entire class of web compatibility complexity that Tauri has...

**Reply by sagudev** [reply]
As for different engines problem, TypeScript can be compiled to specific version of JavaScript and if I remember correctly it was observed that minimum JavaScript version across Tauri platforms is ECMA2018.

But I did encounter that some things were working in Linux but not on Windows so targeting specific JavaScript version isn't complete solution, but definitely part of it.

**Reply by NathanaelA** [reply]
There is also actual differences between the JS in JSCore and V8, some features work, and some features act a bit differently.   For example things like this were broken the last time I check in JSCore, but worked fine in v8.    `let dt = new Date('2019-01-01Z00:00:00');` there are lots of little subtle differences that you might have to deal with as a web developer, but as an app developer for desktop you probably don't want to deal with...   A nice list is to compare Webkit/Safari (JSCore) to Chrome (v8) here: https://kangax.github.io/compat-table/es2016plus/

**Comment by MikeChongCan** [comment]
@NathanaelA 

> memory footprint almost the same as Electron

under what kind of use case? AFAIK, Tauri does not embed node.js so the memory usage should be much smaller in general.

**Reply by NathanaelA** [reply]
See the blog, it answers most the questions but in summry: 
1. You are 100% correct, Tauri doesn't embed node/blink. 
2. It does however use the built in OS WebView which at least under two of the operating systems I ran tests under had virtually the same memory footprint as Electron in anything more complex than a simple "Hello World" :grinning:.   It is harder to measure, just because their is large chunk of shared memory that "hides" memory usage as part of the OS.   When your pages get more complex, it is easier to see the impact on System memory being allocated by the webview.   Which is what most reviews (and afaik, Tauri's doc's don't mention this "memory" as being used by Tauri) so effectively Tauri cheats on its memory usage because its larges chunk of memory usage is effectively hidden elsewhere...      But there are ways to measure it -- you can get a pretty good idea how much "system" memory that the system Webview is allocating for a Tauri application using several techniques.    This "hidden" memory is still used because Tauri needs to display html/css and run JS, it is just the OS is running it on behalf of Tauri and is required for Tauri to actually run.     

The hello world app, Tauri is quite a bit smaller as long as you don't isolate it using AppImage (snap, flatpak, etc) where certain shared memory is no longer shared.   But in a much more detailed app with multiple images -- Tauri is actually within spitting distance of Electron even as a stand alone (non-appimage, using fully system memory 100% shared) app.    See the attached table from by blog...

![image](https://user-images.githubusercontent.com/850871/168699638-ba068ac7-6181-4165-bb2d-06d385c883b6.png)

**Comment by rustdesk** [comment]
https://www.reddit.com/r/FlutterDev/comments/uqr18d/tauri_or_flutter_for_rustdesk_desktop/i8uetbj/?context=3

> With the official support of macOS, I took some time looking at some flutter apps today, and immediately spotted that the caret is forcing a refresh of the whole window. I would consider this a sev1, unfortunately the issue was opened a year ago, still not fixed and marked as a p4 https://github.com/flutter/flutter/issues/85781. This issue alone would make Flutter unsuitable for many of the apps that I would like to write.

**Reply by NathanaelA** [reply]
I don't disagree, there is a reason I don't recommend Flutter normally.  ;-)    But when you are limiting it to only Flutter vs Tauri, it is an easy choice imho based on my experience with both platforms...

**Reply by jimmycoyne** [reply]
Looks like the full window redraw can be solved with a repaint boundary.

**Comment by ethanblake4** [comment]
Fyi, the up-to-date issue for "caret forcing a redraw" is https://github.com/flutter/flutter/issues/33939 and it's already fixed for Android and iOS and was specifically called out in the Flutter 3 announcement blog. Seems likely to be fixed on desktop within a month or two.

**Comment by SoLongAndThanksForAllThePizza** [comment]
@nothingismagick I'm migrating RustDesk from Sciter to Tauri.
Currently, when rendering the remote desktop, RustDesk directly passes the frame buffer to Sciter by calling the render_frame() function. In Tauri, I can render the frame on the canvas, but I can't find an efficient way to pass the frame buffer to the WebView environment for the canvas to render. I can't use WebRTC either because it's not supported on Linux.
Is there any suggestion for this issue? Thank you!

**Reply by zuoxinyu** [reply]
Encountered the same problem when I was writing an electron app many yeas ago. At last we chose shared memory and WebGL to improve the transport/renderer preformance. Copying frame data is unacceptable.

**Reply by zuoxinyu** [reply]
Besides the canvas, I think the [MSE](https://developer.mozilla.org/zh-CN/docs/Web/API/Media_Source_Extensions_API) may be another way to render frames. And it's more closer to the implementation of `render_frame()` in sciter.

**Reply by SoLongAndThanksForAllThePizza** [reply]
Yes, I was trying to find out how to use shared memory in Tauri, but sadly I haven't found any related API.
Also thanks for mentioning MSE, I wasn't considering it because it's still an experimental feature, but after looking into the documentation, it seems that it's well supported. Should definitely give it a try.

**Reply by rustdesk** [reply]
https://github.com/tauri-apps/tauri/discussions/4177#discussioncomment-2788751 from Tauri

**Reply by SoLongAndThanksForAllThePizza** [reply]
Thanks, since it's confirmed that zero-copy is a no-go, I'll try different workarounds and do some benchmarks.

**Reply by zuoxinyu** [reply]
Passing undecoded packet data between webview and backend can significantly reduce the copying consumption, but I'm not sure it works with vpx encoding.

**Reply by SoLongAndThanksForAllThePizza** [reply]
It's supported, see https://www.w3.org/TR/mse-byte-stream-format-registry/.
This is what I'm trying to do(since sending the entire frame buffer to WebView is expensive).

**Reply by zuoxinyu** [reply]
Btw, how do you handle the existing frontend codes?  IMHO, migrating js/css of sciter would be full of pain, did you rewrite those in standard js? 
Actually I tried to migrate it to Tauri two days ago untill saw your comment here.  I'm glad to do some help (since I'm free now), but I didn't find any public reposities in your profile.

**Reply by rustdesk** [reply]
> standard js

Check out this, https://github.com/rustdesk/rustdesk/tree/sciterjs/src/ui, this is standard js (react). We have done some job from sciter.tis to sciter.js before, sciter.js is based on quickjs, I guess it is standard js.

**Reply by rustdesk** [reply]
> any public reposities in your profile

@SoLongAndThanksForAllThePizza is member of us, he will work on the migrating project.

**Reply by zuoxinyu** [reply]
my bad, didn't even realize this is standard js 😆 , looks a bit magical.
```js
    ["on click at div.send span"](evt) {
        this.send();
        view.focus = this.$("input");
    }
```

**Reply by rustdesk** [reply]
@Heap-Hop

**Reply by rustdesk** [reply]
> my bad, didn't even realize this is standard js 😆 , looks a bit magical.
> 
> ```js
>     ["on click at div.send span"](evt) {
>         this.send();
>         view.focus = this.$("input");
>     }
> ```

Grammatically it is, but I think it can not work.

**Reply by SoLongAndThanksForAllThePizza** [reply]
The sciter.js branch can be used as a good reference, but yes, we do need to rewrite a decent amount of code to migrate RustDesk from Sciter to Tauri 😸
I just started the migrating process yesterday, so I'm currently focusing on the main problem which is rendering the frames.
Personally, I think it's necessary to have some discussion and decide on the front-end tech stack before starting to rewrite the front-end part.

**Reply by rustdesk** [reply]
Agree

**Reply by zuoxinyu** [reply]
I recommend to avoid using any framework... that's saying, vanilla js is enough to draw such simple UI.

**Reply by rustdesk** [reply]
UI framework is not necessary, but I think we should use something like this, https://github.com/n8jadams/tauri-typescript-react-boilerplate

**Reply by rustdesk** [reply]
To make our code more readable and maintainable.

**Reply by Heap-Hop** [reply]
> To make our code more readable and maintainable.

Yes, I think typescript is necessary, hope you start with typescript. @SoLongAndThanksForAllThePizza

**Reply by SoLongAndThanksForAllThePizza** [reply]
Using MSE to display VPX encoded frames is proved possible(I made a working demo). No existing libraries can satisfy our use case, so I had to manually write a binary encoder(more troublesome than I thought!). Since this problem is solved, I can start rewriting the front-end part. I have some other deadlines to meet, so I think I'll start next Monday.
I prefer to use TS over JS in the first place so I'm glad to hear that!
Let me make a conclusion:
1. No UI Framework for now.
2. We use Typescript.
3. I'll choose Vite as the building tool for the front-end development. Can't live without HMR 😸 
Let me know if you agree with the choice above. Thank you!

**Reply by rustdesk** [reply]
Agree

**Reply by rustdesk** [reply]
Due to tauri's worse ipc performance and MSE's bad support of 265 decoding etc, we may have to give up Tauri. @nothingismagick

**Reply by rustdesk** [reply]
@SoLongAndThanksForAllThePizza  @Heap-Hop  @zuoxinyu

**Reply by LifeIsStrange** [reply]
@rustdesk seems logical to try Electron next then

**Reply by rustdesk** [reply]
> Electron

But a single small portable exe is an important feature for us.

**Reply by sagudev** [reply]
Flutter_rust_bridge has ZeroCopyBuffer for Vec<u8>, which should allow fast image sending. http://cjycode.com/flutter_rust_bridge/feature/zero_copy.html

**Reply by LifeIsStrange** [reply]
@rustdesk I don't think a 27MB difference is really meaningful over the other criteria but that's just my point of view also that's just for appimage, maybe it would be different on flatpack or snap?

**Reply by zuoxinyu** [reply]
As for flutter I can't give any suggestions, since I'm not familiar with it and personally I don't trust google's stack.

Technically, all webview-based framework would face the same critical frame rendering problom. There is an evil, final workaround: using native window to render the video, and put a (maybe offscreen rendered) webview layer on top of the native window for user controls.

And, personally again, keep using sciter is another reasonable choice.

**Reply by zuoxinyu** [reply]
I found there is `yuv.wasm` in mobile codes, if WASM is an acceptable way, I think it can solve the h265 problem, but without hardware acceleration.

**Reply by SoLongAndThanksForAllThePizza** [reply]
Another perspective: If RustDesk Web Client is considered important, developing a Tauri client ≈ developing a web client.
I'm not an expert in Flutter so tell me if I'm wrong, but I think native H5+JS should give a better user experience compared with Flutter Web.
If a web client is not considered an important feature for RustDesk, then simply using Flutter Web should be enough.

**Reply by rustdesk** [reply]
>  `yuv.wasm` 

super slow

**Reply by LifeIsStrange** [reply]
@zuoxinyu electron can support h265, you "just" need to tweak the ffmpeg build flags
https://github.com/AAAhs/electron-hevc

**Reply by SoLongAndThanksForAllThePizza** [reply]
Using WASM to decode without using hardware acceleration is unacceptably slow. I think based on what we've already discussed, Flutter is a better choice. Relying on MSE will be a huge restriction.

**Reply by zuoxinyu** [reply]
> Using WASM to decode without using hardware acceleration is unacceptably slow. I think based on what we've already discussed, Flutter is a better choice. Relying on MSE will be a huge restriction.

Yes, extremely slow, can't even afford a 2k stream.  Should be definitely out of choices.

**Reply by nothingismagick** [reply]
We are planning on releasing an egui feature which is a tao window immediately after the 1.0 stable release. If you would like an early preview of it, feel free to DM me on our discord: @nothingismagick is my name there.

**Reply by rustdesk** [reply]
> egui

But as I know, egui has problem with non-latin characters. Will you have the same problem? https://github.com/emilk/egui#faq

**Reply by jimmycoyne** [reply]
Since you’re already using flutter for mobile, does it not make sense to evaluate that stack on desktop? If (and I know it’s a big if) it works out, having the common stack on both will really help regarding development speed and ability for community to contribute to all platforms ?

**Reply by rustdesk** [reply]
@zuoxinyu Do you still have interest if we change to Flutter?

**Reply by zuoxinyu** [reply]
Then I'm not going to touch the UI part... maybe I can focus on the backend.

**Reply by nothingismagick** [reply]
The only problem with non-latin characters is supplying a font as i read that FAQ

**Reply by zuoxinyu** [reply]
> personally I don't trust google's stack.

https://www.reddit.com/r/technology/comments/1cgivld/google_lays_off_staff_from_flutter_dart_and/

recall.

**Comment by kawpii** [comment]
Try [Compose for Desktop](https://www.jetbrains.com/lp/compose-mpp/) too

**Comment by trymeouteh** [comment]
Both are good but Flutter seems to be better in my opinion since it natively compiles on every platform and Flutter/Dart is growing.

I always prefer if apps that are design for multiple devices to be written in Flutter.

**Reply by LifeIsStrange** [reply]
You know that natively compiles doesn't mean anything right? Dart AOT is slower than its JIT mode and being slow is not a virtue. As for the flutter SDK and tauri/electron/ionic they are all written in the same language: C++.
Moreover flutter use the chrome 2D rendering engine: Skia.
The flutter/dart ecosystem is order of magnitudes smaller than the web ecosystem and therefore lacks many libraries in all metrics (performance, featureness, security audits)

**Reply by ethanblake4** [reply]
@LifeIsStrange Dart AOT was slower than it's JIT mode years ago, but it's long since surpassed it (at least in the vast majority of cases). And JIT is not allowed on iOS so it's definitely good that Flutter doesn't use it.

**Comment by mon-jai** [comment]
Is there any ETA or beta release for the flutter rewritten version of the client?

**Comment by yguel** [comment]
In a similar discussion with a dev team, someone brought the fact that with Tauri you can use webgl whereas in Flutter the opengl support is very poor.
The dev team uses Tauri for that reason and is very happy with it. However they target web and desktop, not a mobile platform.

We have the need to also target mobile platforms for the interface of our new project. Does someone has experience with multi-platform development requiring opengl ?

At some point in our discussion with the dev team someone also suggested to develop inside Unity or another game engine supporting multi-platforms.
As our application backend is in rust, this would require bindings, but this is also an interesting alternative not explored by @NathanaelA

**Reply by nothingismagick** [reply]
Tauri will soon be offering Mobile (technically already available on our `next` branch.)

**Reply by yguel** [reply]
We have to test it with some code using BLE already written for android and iOS.

**Reply by NathanaelA** [reply]
My understanding is with Flutter OpenGL/Vulkan/Metal support is actually very good, since that is actually how the engine works all the time  (Everything is driven via the native graphics sub-system).   There are even a couple Game engine projects written in Flutter/Dart.    But I haven't personally tried calling opengl directly myself, but you might want to try it out before dismissing it.   

As for using Unity (very buggy) or Unreal (very heavy), I have used them.  :grinning:      Actually currently investigating [Sciter](https://sciter.com/) and a couple other not quite as mainstream mobile/desktop engines...   ;-)

I'll have to look into Tauri's mobile stuff, but I suspect it will be very much like Ionic/PhoneGap/Cordova with the same pro's and con's  if it is the same binding to native browser system as they do on the Desktop....

**Comment by SakhnevichKirill** [comment]
If someone interested, i'm starting migration sciter -> tauri + react. 
Welcome to contribute:
https://github.com/SakhnevichKirill/rustdesk_tauri

**Reply by mon-jai** [reply]
They have already migrated Rustdesk to Flutter.

Couldn't we just compile the app with Flutter web to support web platforms?

**Reply by liudonghua123** [reply]
The code in https://github.com/SakhnevichKirill/rustdesk_tauri seems not complete?

**Reply by SakhnevichKirill** [reply]
> They have already migrated Rustdesk to Flutter.
> 
> Couldn't we just compile the app with Flutter web to support web platforms?

I haven't worked with Flutter, and I'm currently studying React and Tauri. I'm interested in working with Rastdesk, so I wanted to implement this. I noticed that many people are interested in implementing on Tauri, so maybe it will be useful to someone. Of course, development on Tauri and React should not interfere with development on Flutter

**Reply by SakhnevichKirill** [reply]
> The code in https://github.com/SakhnevichKirill/rustdesk_tauri seems not complete?

Yes, the development is at an early stage. My main focus is the backend development and I am familiar with react. Сurrent task is to reproduce frontend calls to Rust through Tauri commands

**Comment by condominiolosnogalesph-hub** [comment]
gracias
